docs/proposals/31-inplace-node-update.md

```diff
@@ -0,0 +1,441 @@
+
+This design was chosen to leverage the `machine-controller-manager`'s node rollout orchestration functionality while ensuring that it remains a standalone component.
+
+### Prerequisites
+
+To perform an in-place OS update, the following prerequisites must be met:
+
+- The node's machine image or operating system must support in-place updates, providing a tool or utility to initiate the update process.
+- The update mechanism should ensure reliability by:
+  - Booting into the new version upon a successful update.
+  - Falling back to the previous version in case of failure.
+
+Additionally, the update tool or utility may offer configuration options, such as:
+
+- Retry Configuration: The ability to define the number of retries for the update process.
+- Registry Configuration: The option to specify the registry from which the OS image is pulled.
+
+### Update Strategies
+
+Gardener will introduce two additional update strategies, `AutoInPlaceUpdate` and `ManualInPlaceUpdate`, for worker pools, in addition to the existing rolling update which will be now named as `AutoReplaceUpdate` strategy. In the future, there could also be `ManualReplaceUpdate` which could give more control to the users in rolling updates as well. The two new strategies are described as follows:
+
+#### `AutoInPlaceUpdate` strategy
+
+1. On a shoot spec change for a worker pool, `machine-controller-manager` will label all nodes of the updated worker pool with the `node.machine.sapcloud.io/candidate-for-update` label. This label is used to identify nodes that require an update.
+2. `machine-controller-manager` will add the necessary annotation to the nodes to prevent them from being scaled down by [`cluster-autoscaler`](https://github.com/gardener/autoscaler/tree/machine-controller-manager-provider/cluster-autoscaler) during the update process.
+3. `machine-controller-manager` will select as many nodes as are permitted by the `maxUnavailable` setting for the worker pool and label it/them with `node.machine.sapcloud.io/selected-for-update`. This label can be used to identify nodes that are about to or are currently undergoing an update.
+4. `machine-controller-manager` will cordon and drain the node/nodes and label it/them with the `node.machine.sapcloud.io/ready-for-update` label once the drain is completed.
+5. `gardener-node-agent` will detect the `node.machine.sapcloud.io/ready-for-update` label and perform the update on the machine. It will wait for the update to complete (with a specified timeout).
+6. Once the machine is updated, `gardener-node-agent` will delete all the pods in the node so that they will get recreated and the node will be labelled by the `gardener-node-agent` with the `node.machine.sapcloud.io/update-successful` label.
+7. `machine-controller-manager` will `uncordon` any node with the `node.machine.sapcloud.io/update-successful` label and remove this and all other update-related labels. Once the node becomes `Ready`, it can host workload again.
```

Comment:
If the update to a newer version is not successful and the update tool has reverted the OS version to the older version (that existed prior to the update) then its still effectively functional. Should it remain `cordoned` and unavailable as a target for the kube-scheduler?

Reply:
Yes, let's make it so. otherwise, we will have to go through the drain again. Anyways, if the failed nodes reach `maxUnvailable`, MCM stops the update. This is also what happens in the case of normal updates as well, right?

Reply:
ok